server/src/main/java/com/angelomelonas/grpcwebchat/common/ChatSession.java

```diff
@@ -0,0 +1,64 @@
+package com.angelomelonas.grpcwebchat.common;
+
+import com.angelomelonas.grpcwebchat.Chat.Message;
+import io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+
+public class ChatSession {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ChatSession.class);
+
+    private String username;
+    private StreamObserver<Message> responseObserver;
+
+    private boolean isSubscribed;
+
+    public ChatSession() {
+        this.isSubscribed = false;
+    }
+
+    public void subscribe(String username, StreamObserver responseObserver) {
+        this.username = username;
+        this.responseObserver = responseObserver;
+        this.isSubscribed = true;
+
+        LOGGER.info("Client with username {} subscribed.", this.username);
+    }
+
+    public void unsubscribe() {
+        if (!this.isSubscribed) {
+            throw new IllegalArgumentException("Cannot unsubscribe. Session not subscribed.");
+        }
+
+        try {
+            // Try to close the stream.
+            this.responseObserver.onCompleted();
+        } catch (IllegalStateException exception) {
+            throw exception;
```

Comment:
You are just throwing the exception anyways.
Maybe you need to log something instead and set `isSubscribed = false` ?